### ISSUE TITLE
Questions Regarding Model and Loss Structure in Your Paper and Code

### ISSUE THREAD

**Issue by Olrun** [NJU-RL/ACORM#7]
Hello,

First of all, I would like to express my appreciation for your work. I have read your paper with great interest and have been exploring the code that accompanies it. During my review, I encountered a few questions regarding the model structure and the learning process, which I hope you can clarify.

### 1. Agent Embedding Part

**1.1** It appears that the Q-value and the Role representation are generated by separate networks, rather than being produced by a single network as described in the paper.
```
def choose_action(self, obs_n, last_onehot_a_n, role_embedding, avail_a_n, epsilon):
        with torch.no_grad():
            if np.random.uniform() < epsilon:  # epsilon-greedy
                # Only available actions can be chosen
                a_n = [np.random.choice(np.nonzero(avail_a)[0]) for avail_a in avail_a_n]
            else:
                inputs = copy.deepcopy(obs_n)
                if self.add_last_action:
                    inputs = np.hstack((inputs, last_onehot_a_n))
                inputs = np.hstack((inputs, role_embedding.to('cpu')))
                inputs = torch.tensor(inputs, dtype=torch.float32)
                inputs = inputs.to(self.device)

                q_value = self.eval_Q_net(inputs)
                avail_a_n = torch.tensor(avail_a_n, dtype=torch.float32)  # avail_a_n.shape=(N, action_dim)
                q_value = q_value.to('cpu')
                q_value[avail_a_n == 0] = -float('inf')  # Mask the unavailable actions

                a_n = q_value.argmax(dim=-1).numpy()
        return a_n
```
**1.2** In Figure 1 of the paper, the Q-network is described as taking the last_action and observation as inputs. However, I noticed that in the code, the Q-network also takes the Role representation (z_t) as an additional input. Could you clarify this discrepancy?

### 2. Learning Part

**2.1** I came across the term pretrain_agent_embedding in the runner code. However, I could not find any reference to this loss function in your paper. Could you explain its role and provide more details?
```
      if self.agent_embed_pretrain_epoch < self.args.agent_embed_pretrain_epochs:
                if self.replay_buffer.current_size >= self.args.batch_size:
                    self.agent_embed_pretrain_epoch += 1
                    agent_embedding_loss = self.agent_n.pretrain_agent_embedding(self.replay_buffer)
                    self.pretrain_agent_embed_loss.append(agent_embedding_loss.item())
```
**2.2** The pseudo-code in your paper suggests that contrastive learning and QMix are conducted concurrently. However, the code seems to perform contrastive learning exclusively for the first 120 episodes, followed by QMix-based training without further contrastive learning. Is this intentional, and if so, could you elaborate on the rationale behind this approach?

**2.3** I also noticed that in the code, only the steps taken during QMix training are added to the total step count, whereas the pretraining steps (contrastive learning and agent embedding learning) are not included. Could you confirm if this is the intended behavior?
```
        while self.total_steps < self.args.max_train_steps:
            if self.total_steps // self.args.evaluate_freq > evaluate_num:
                self.evaluate_policy()  # Evaluate the policy every 'evaluate_freq' steps
                evaluate_num += 1

            _, _, episode_steps = self.run_episode_smac(evaluate=False)  # Run an episode
            
            if self.agent_embed_pretrain_epoch < self.args.agent_embed_pretrain_epochs:
                if self.replay_buffer.current_size >= self.args.batch_size:
                    self.agent_embed_pretrain_epoch += 1
                    agent_embedding_loss = self.agent_n.pretrain_agent_embedding(self.replay_buffer)
                    self.pretrain_agent_embed_loss.append(agent_embedding_loss.item())
            else:
                if self.recl_pretrain_epoch < self.args.recl_pretrain_epochs:
                    self.recl_pretrain_epoch += 1
                    recl_loss = self.agent_n.pretrain_recl(self.replay_buffer)
                    self.pretrain_recl_loss.append(recl_loss.item())
                    
                else:                                                          
                    self.total_steps += episode_steps
                    if self.replay_buffer.current_size >= self.args.batch_size:
                        self.agent_n.train(self.replay_buffer)  # Training
```

I would greatly appreciate your insights on these points. Thank you for your time and for sharing your work with the community.

**Comment by huzican** [comment]
I apologize for not responding promptly as I was preoccupied with other commitments.
1. The structure resembles a residual connection, where the role embeddings are computed from the observation and action embeddings, and Q is jointly determined by the role along with both observation and action.
2. a)  The pretraining phase was originally introduced due to concerns that imprecise agent embeddings might result in inaccurate role clustering, potentially compromising the stability of the training pipeline. However, subsequent experiments revealed that the framework could successfully bootstrap without pretraining, making it feasible to set the pretrain steps to zero.
b) This modification was motivated by our observation that roles exhibited clear clustering patterns in the later training phases. To optimize computational efficiency, we decreased the frequency of contrastive learning updates. Notably, maintaining the original update frequency showed no substantial effect on the learning trajectories.
c）Our analysis of the pretrain steps revealed that its computational overhead is insignificant in relation to the total training duration.

**Comment by Olrun** [comment]
Thank you for your clear and insightful response. I look forward to seeing more interesting research from you in the future.
